opm/material/fluidsystems/BlackOilFluidSystem.hpp

```diff
@@ -313,12 +313,12 @@ class BlackOil : public BaseFluidSystem<Scalar, BlackOil<Scalar> >
         switch (phaseIdx) {
         case waterPhaseIdx: return waterDensity<LhsEval>(T, p, regionIdx);
         case gasPhaseIdx: {
-            const auto& XgO = FsToolbox::template toLhs<LhsEval>(fluidState.massFraction(gasPhaseIdx, oilCompIdx));
-            return gasDensity<LhsEval>(T, p, XgO, regionIdx);
+            const auto& Rv = getRv_<LhsEval>(fluidState, regionIdx);
+            return gasDensity<LhsEval>(T, p, Rv, regionIdx);
         }
         case oilPhaseIdx: {
-            const auto& XoG = FsToolbox::template toLhs<LhsEval>(fluidState.massFraction(oilPhaseIdx, gasCompIdx));
-            return oilDensity<LhsEval>(T, p, XoG, regionIdx);
+            const auto& Rs = getRs_<LhsEval>(fluidState, regionIdx);
```

Comment:
the non-change is because of the non-blackoil models: the basic intention is to keep the fluid system API as it is (because the semantics `R_s` and `R_v` are just rubbish in the general case, and even for the black-oil model their definition is quite awkward IMO), but to introduce a `BlackOilFluidState` class that features methods that directly return `R_s` and `R_v`. This PR does not yet do this, but I envision this in the medium term. with this, the `getR[sv]_()` methods can be specialized for `BlackOilFluidState`, i.e., the fluid system can still be used with generic fluid states, but it is slightly faster if the model uses `BlackOilFluidState`.